### PR TITLE
[rustdoc] add sub settings

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
This PR is to give a finer control over what types are automatically expanded or not as well as the possibility to add sub-settings in the settings page.

Screenshot from 2019-09-23 00-46-14

r? @Mark-Simulacrum

### CODE REVIEW

**Comment by fbstj** [comment]
is this just a reopening of #60778 ?

**Comment by GuillaumeGomez** [comment]
No, I made some changes. For instance, there is no more global checkbox, just a title for sub-settings.

**Comment by Mark-Simulacrum** [comment]
So to be clear, this PR changes the current "Auto-hide item declarations." into a set of preferences for each item type?

Can you update the top-level comment with an answer to this question?
 * How does this work when the previously loaded version of the page had auto-hide item declaration checked (and unchecked)? What about the reverse (i.e., going from the new version to the old)?
 * Why were these particular defaults chosen?

Personally, I never really found it all that useful to have a global hide setting for item declarations here so I would not want this feature -- I do generally work on at least a 15" screen though so in that sense I have relatively large screen real estate. Is this feature wanted by folks? And in the "yes I'd use this" not the "well, why not" sense?

**Comment by GuillaumeGomez** [comment]
> How does this work when the previously loaded version of the page had auto-hide item declaration checked (and unchecked)? What about the reverse (i.e., going from the new version to the old)?

No back compatibility if this is your question.

> Why were these particular defaults chosen?

Debated on the previously opened PR.

> Personally, I never really found it all that useful to have a global hide setting for item declarations here so I would not want this feature -- I do generally work on at least a 15" screen though so in that sense I have relatively large screen real estate. Is this feature wanted by folks? And in the "yes I'd use this" not the "well, why not" sense?

I've been asked to do this a lot of times actually (IRL and IRC). I personally don't care much about this feature but people seems to want to have a better control over the types declaration view.

**Comment by Mark-Simulacrum** [comment]
> No back compatibility if this is your question.

But, uh, what happens? Presumably we don't just error out and not run at all...

> Debated on the previously opened PR.

Do you have a link? I don't see any such discussion on https://github.com/rust-lang/rust/pull/60778...

> I've been asked to do this a lot of times actually (IRL and IRC). I personally don't care much about this feature but people seems to want to have a better control over the types declaration view.

I would personally echo @QuietMisdreavus's concerns on https://github.com/rust-lang/rust/pull/60778 that this does not seem to really carry it's weight. However, if there is indeed lots of demand, it also seems like a 'not bad' idea.

**Comment by GuillaumeGomez** [comment]
> But, uh, what happens? Presumably we don't just error out and not run at all...

It doesn't take it into account. That's pretty much it.

> Do you have a link? I don't see any such discussion on #60778...

Seems like it wasn't there... Multi-channels communications are bad...

> I would personally echo @QuietMisdreavus's concerns on #60778 that this does not seem to really carry it's weight. However, if there is indeed lots of demand, it also seems like a 'not bad' idea.

Like I said, I didn't do it just because. People asked me for it and at some point I just said ok and did it.

**Comment by Mark-Simulacrum** [comment]
I'm going to r? @QuietMisdreavus here as a rustdoc team member, I don't personally feel comfortable driving this

**Comment by JohnCSimon** [comment]
Ping from triage -
@QuietMisdreavus  this seems to be waiting on a review
Thanks.

**Comment by ProgrammaticNajel** [comment]
Ping from triage. @QuietMisdreavus any updates on this? Thanks.

**Comment by JohnCSimon** [comment]
Pinging from triage:
 @QuietMisdreavus can you please review this PR?
Also @GuillaumeGomez can you please resolve the merge conflict?
Thank you!

**Comment by GuillaumeGomez** [comment]
Rebased.

**Comment by GuillaumeGomez** [comment]
Fixed the issue.

cc @QuietMisdreavus

**Comment by JohnCSimon** [comment]
Ping from triage 
@QuietMisdreavus can you please review this PR?
Thanks

File: src/librustdoc/html/static/settings.css [hunk 1]
```diff
@@ -59,3 +67,9 @@ input:checked + .slider:before {
 	-ms-transform: translateX(19px);
 	transform: translateX(19px);
 }
+
+.setting-line > .sub-setting {
```
Comment (kinnison) [inline src/librustdoc/html/static/settings.css]: To match the change to `sub-settings` you'd need:
```suggestion
.setting-line > .sub-settings {
```

**Review by kinnison — CHANGES_REQUESTED** [review]
In addition to the pluralisation comment I have some concerns that we are not testing the generation of this content at all.

It would probably behoove us to ensure there's at least a basic test checking for the presence of the settings in the generated output.

**Comment by GuillaumeGomez** [comment]
We have a UI server but the integration with highfive is buggy and needs to be fixed. (Maybe when @pietroalbini has some time, we could take a look to understand what's going on)

**Comment by GuillaumeGomez** [comment]
Updated.

**Review by kinnison — APPROVED** [review]
LGTM :+1:

**Comment by kinnison** [comment]
Regarding some basic regression testing, could we do a basic rustdoc-ui test using `htmldocck.py` ?

**Comment by GuillaumeGomez** [comment]
It allows to check the DOM, so it depends what you want to check?

**Comment by kinnison** [comment]
I was just thinking that a basic check demonstrating that the settings have made it out in a form we were expecting -- that way if the shape is changed in the future, there's a regression test reminding people to re-check it and update the test if it's good.  I wouldn't block a merge on that but you were saying you wanted more tests :D

**Comment by GuillaumeGomez** [comment]
It's more UI tests then. It's scheduled but I need to have highfive fully working on this so we don't have bad surprises.

**Comment by kinnison** [comment]
Okay, if you want to make a note (issue?) for the need for the test, and merge the fix in the meantime, I'm good with that.

**Comment by GuillaumeGomez** [comment]
Done [here](https://github.com/rust-lang/rust/issues/66181). Thanks for your review!

@bors: r=kinnison

**Pull request merged**